### PR TITLE
Add multi-arch support for clamav scan

### PR DESCRIPTION
**Pull request by yashvardhannanavati**
[CVP-4040] clamav scan should identify an image index, get all the image manifests in it and scan each one of them.

# Before you complete this pull request ...

Look for any open pull requests in the repository with the title "e2e-tests update" and 
see if there are recent e2e-tests updates that will be applicable to your change.

### CODE REVIEW

**Comment by yashvardhannanavati** [comment]
single-arch & multi-arch test
clamav-scan-multi-arch
clamav-scan-single-arch

File: task/clamav-scan/0.1/clamav-scan.yaml [hunk 1]
```diff
@@ -63,40 +65,101 @@ spec:
...
+        jq -s -rce '
+          reduce .[] as $item ({"timestamp":"0","namespace":"","successes":0,"failures":0,"warnings":0,"result":"SUCCESS","note":"All checks passed successfully"};
+            {
+            "timestamp" : (if .timestamp < $item.timestamp then $item.timestamp else .timestamp end),
+            "namespace" : $item.namespace,
+            "successes" : (.successes + $item.successes),
+            "failures" : (.failures + $item.failures),
+            "warnings" : (.warnings + $item.warnings),
+            "result" : (if $item.result != "SUCCESS" then $item.result else .result end),
+            "note" : (if $item.result != "SUCCESS" then $item.note else .note end)
```
Comment (yashvardhannanavati) [inline task/clamav-scan/0.1/clamav-scan.yaml:156]: This part of the result needs to be refined. Currently, it returns the `result` and `note` of the scan result that is not `SUCCESS` but as mentioned in https://github.com/redhat-appstudio/architecture/blob/main/ADR/0030-tekton-results-naming-convention.md#test_output-schema-validation, the result can be any of `["SUCCESS", "FAILURE", "WARNING", "SKIPPED", "ERROR"]`. So in an aggregated result, we probably need to rank the result statuses. I do not see that currently in the ADR.

As of today, the current approach may work for multi-arch images that build for two arches but the approach is not robust for builds supporting 3 or more arches. @arewm @ralphbean @dirgim @MartinBasti @lcarva Can I please request you to share your thoughts on this?

File: task/clamav-scan/0.1/clamav-scan.yaml [hunk 2]
```diff
@@ -63,40 +65,101 @@ spec:
...
+            })' /tekton/home/clamscan-ec-test-*.json | tee $(results.TEST_OUTPUT.path)
+
+        # Add the IMAGE_DIGEST to IMAGES_PROCESSED result string anyways because even if the image index itself is not scanned,
+        # it gets resolved to amd64 arch which has been scanned
+        digests_processed+="${digests_processed:+, }\"$IMAGE_DIGEST\""
+        echo "${images_processed_template/\[%s]/[$digests_processed]}" | tee $(results.IMAGES_PROCESSED.path)
```
Comment (tkdchen) [inline task/clamav-scan/0.1/clamav-scan.yaml:162]: There is a good practice that script references tekton variable via environment variable.

Comment (MartinBasti) [reply task/clamav-scan/0.1/clamav-scan.yaml:162]: that's for tekton parameters if I'm not mistaken

File: task/clamav-scan/0.1/clamav-scan.yaml [hunk 3]
```diff
@@ -63,40 +65,101 @@ spec:
...
+            echo "Scanning image for arch $arch. This operation may take a while."
+            clamscan $destination -ri --max-scansize=4095M --max-filesize=4095M \
+              --max-scantime=0 --max-files=0 --max-recursion=1000 --max-dir-recursion=20000 --max-embeddedpe=4095M \
+              --max-htmlnormalize=10M --max-htmlnotags=4095M --max-scriptnormalize=5M --max-ziptypercg=4095M \
+              --max-partitions=50000 --max-iconspe=100000 --max-rechwp3=20000 --pcre-match-limit=100000000 --pcre-recmatch-limit=2000000 \
+              --pcre-max-filesize=4095M --alert-exceeds-max=yes \
+              --alert-encrypted=yes --alert-encrypted-archive=yes --alert-encrypted-doc=yes --alert-macros=yes \
+              --alert-phishing-ssl=yes --alert-phishing-cloak=yes --alert-partition-intersection=yes \
+              | tee /tekton/home/clamscan-result-$arch.log || true
+            echo "Executed-on: Scan was executed on version - $(clamscan --version)" | tee -a /tekton/home/clamscan-result-$arch.log
```
Comment (MartinBasti) [inline task/clamav-scan/0.1/clamav-scan.yaml:99]: Can we avoid this duplication?

File: task/clamav-scan/0.1/clamav-scan.yaml (hunk 1 above)
Comment (arewm) [reply task/clamav-scan/0.1/clamav-scan.yaml:156]: Great catch!

File: task/clamav-scan/0.1/clamav-scan.yaml (hunk 3 above)
Comment (yashvardhannanavati) [reply task/clamav-scan/0.1/clamav-scan.yaml:99]: yes! With the new hacbs-test image, we can get rid of this. I am working on it and will update it by EOD

Comment (dirgim) [reply task/clamav-scan/0.1/clamav-scan.yaml:99]: Resolving this convo as it has been addressed.

File: task/clamav-scan/0.1/clamav-scan.yaml [hunk 4]
```diff
@@ -63,40 +65,78 @@ spec:
...
         cd content
-        echo Extracting image.
-        if ! oc image extract --registry-config ~/.docker/config.json $imageanddigest; then
-          echo "Unable to extract image. Skipping ClamAV scan!"
-          exit 0
-        fi
+        echo "Extracting image(s)."
+
+        # Get the arch and image manifests by inspecting the image. This is mainly for identifying image indexes
+        image_manifests=$(get_image_manifests -i ${imageanddigest})
```
Comment (jsztuka) [inline task/clamav-scan/0.1/clamav-scan.yaml:76]: Can't find definition of `get_image_manifests` where is this called from? I supose this is native binary?

File: task/clamav-scan/0.1/clamav-scan.yaml [hunk 5]
```diff
@@ -63,40 +65,78 @@ spec:
...
+            "note" : (if .result == "" or ($item.result == "SKIPPED" and .result == "SUCCESS") or ($item.result == "WARNING" and (.result == "SUCCESS" or .result == "SKIPPED")) or ($item.result == "FAILURE" and .result != "ERROR") or $item.result == "ERROR" then $item.note else .note end)
+            })' /tekton/home/clamscan-ec-test-*.json | tee $(results.TEST_OUTPUT.path)
 
-        # workaround: due to a bug in ec-cli, we cannot generate json and appstudio output at the same time, running it again
-        EC_EXPERIMENTAL=1 ec test \
-          --namespace required_checks \
-          --policy /project/clamav/virus-check.rego \
-          -o appstudio \
-          /tekton/home/clamscan-result-log.json | tee $(results.TEST_OUTPUT.path) || true
+        digests_processed_string=$(IFS=,; echo "${digests_processed[*]}")
```
Comment (jsztuka) [inline task/clamav-scan/0.1/clamav-scan.yaml:138]: What does IFS stand for?

**Comment by jsztuka** [comment]
Looks fine, once the e2e tests pass, I will provide +1

**Review by dirgim — APPROVED** [review]
LGTM code-wise. We're expecting #919 to unblock the e2e CI issues.

File: task/clamav-scan/0.1/clamav-scan.yaml (hunk 5 above)
Comment (yashvardhannanavati) [reply task/clamav-scan/0.1/clamav-scan.yaml:138]: It means `Internal Field Separator`. The default is " " for the entire script but here, I am only changing it in the context of this command

**Comment by yashvardhannanavati** [comment]
From the logs it seems like sast-synk-check failed. I'm not sure if the failure is realted to the PR.

**Review by jsztuka — COMMENTED** [review]
lgtm

**Pull request merged**